ethcontract/src/transaction/gas_price.rs

```diff
@@ -1,121 +1,60 @@
-    /// the transaction `SendFuture` to not query the node for a gas price
-    /// estimation.
-    Value(U256),
+    /// Legacy type of transactions, using single gas price value. Equivalent to sending
+    /// eip1559 transaction with max_fee_per_gas = max_priority_fee_per_gas = gas_price
+    Legacy(U256),
+
+    /// Eip1559 type of transactions, using two values (max_fee_per_gas, max_priority_fee_per_gas)
+    Eip1559((U256, U256)),
 }
 
 impl GasPrice {
-    /// A low gas price. Using this may result in long confirmation times for
-    /// transactions, or the transactions not being mined at all.
-    pub fn low() -> Self {
-        GasPrice::Scaled(0.8)
-    }
-
-    /// A high gas price that usually results in faster mining times.
-    /// transactions, or the transactions not being mined at all.
-    pub fn high() -> Self {
-        GasPrice::Scaled(6.0)
-    }
-
-    /// Returns `Some(value)` if the gas price is explicitly specified, `None`
-    /// otherwise.
-    pub fn value(&self) -> Option<U256> {
+    /// Prepares the data for transaction. Returns tuple:
+    /// (gas_price, max_fee_per_gas, max_priority_fee_per_gas, transaction_type)
+    pub fn resolve_for_transaction(
+        &self,
+    ) -> (Option<U256>, Option<U256>, Option<U256>, Option<U64>) {
         match self {
-            GasPrice::Value(value) => Some(*value),
-            _ => None,
+            GasPrice::Legacy(value) => (Some(*value), None, None, None),
+            GasPrice::Eip1559(pair) => (None, Some(pair.0), Some(pair.1), Some(2.into())),
         }
     }
-
-    /// Resolves the gas price into a value. Returns a future that resolves once
-    /// the gas price is calculated as this may require contacting the node for
-    /// gas price estimates in the case of `GasPrice::Standard` and
-    /// `GasPrice::Scaled`.
-    pub async fn resolve<T: Transport>(self, web3: &Web3<T>) -> Result<U256, ExecutionError> {
-        let resolved_gas_price = match self {
-            GasPrice::Standard => web3.eth().gas_price().await?,
-            GasPrice::Scaled(factor) => {
-                let gas_price = web3.eth().gas_price().await?;
-                scale_gas_price(gas_price, factor)
-            }
-            GasPrice::Value(value) => value,
-        };
-
-        Ok(resolved_gas_price)
-    }
-
-    /// Resolves the gas price into an `Option<U256>` intendend to be used by a
-    /// `TransactionRequest`. Note that `TransactionRequest`s gas price default
-    /// to the node's estimate (i.e. `GasPrice::Standard`) when omitted, so this
-    /// allows for a small optimization by foregoing a JSON RPC request.
-    pub async fn resolve_for_transaction_request<T: Transport>(
-        self,
-        web3: &Web3<T>,
-    ) -> Result<Option<U256>, ExecutionError> {
-        let gas_price = match self {
-            GasPrice::Standard => None,
-            _ => Some(self.resolve(web3).await?),
-        };
-
-        Ok(gas_price)
-    }
 }
 
 impl Default for GasPrice {
```

Comment:
Maybe a `Default` gas price implementation doesn't make much sense anymore. I'd be OK to remove it, unless its used somehwere.

Reply:
With erasing `GasPrice::Standard` and `GasPrice::Scaled` we are enforcing a user of `TransactionBuilder` to provide gas price or else no gas price will be used for this transaction (resolving of the gas price by asking the node for estimate is removed from `TransactionBuilder'). Is this the desired behavior? I would say yes based on the: https://github.com/gnosis/ethcontract-rs/pull/638#discussion_r716433621

Reply:
In the past, we had:
```
GasPrice::Standard // use `eth_estaimteGasPrice`
GasPrice::Scaled(..) // use `eth_estaimteGasPrice` and mltiply by factor, wanted to get rid of this
GasPrice::Value(..) // use the specified value.
```

I was more suggesting getting rid of `GasPrice` and instead using `Option<U256>` on the `TransactionBuilder`. In this case, we would have:
```
None // Note that `rust-web3` crate allows `None` values for its transaction requests.
Some(..) // behaves like the `GasPrice::Value(..)` used to
```

With the addition of EIP-1559 gas pricing, I would suggest using `Option<GasPrice>` in the `TrasnactionBuilder` so that:
```
None // Note that `rust-web3` crate allows `None` values for its transaction requests
Some(GasPrice::Legacy(..)) // behaves like the `GasPrice::Value(..)` used to
Some(GasPrice::Eip1559(..)) // the cool new stuff :D
```

Reply:
Yes, with the last commit (https://github.com/gnosis/ethcontract-rs/pull/638/commits/87b0e64b06723718d408a24d61fec44d11bd3407) I think the implementation is complaint with the above comment. Also, moving last comment to PR description for more visibility.